### PR TITLE
 table: index.addindices reuse temp buffer

### PR DESCRIPTION
**Pull request by winkyao**
```
master:
[direct] kv num = 24576000 / kv size = 2724864000
[direct] cost avg = 65.792 sec

this pr:
[direct] kv num = 24576000 / kv size = 2724864000
[direct] cost avg = 50.687 sec
```

about 22.7% improvement.

@coocood @shenli PTAL

### CODE REVIEW

**Comment by shenli** [comment]
Any benchmark result?

File: sessionctx/variable/session.go [hunk 1]
```diff
@@ -109,6 +109,31 @@ func (tc *TransactionContext) ClearDelta() {
 	tc.TableDeltaMap = nil
 }
 
+// InsertValuesBufs is buffers used by insert statement.
```
Comment (shenli) [inline sessionctx/variable/session.go]: Actually, the buffers are not only used in the insert statement. But also in  delete/update/replace statements.

**Comment by winkyao** [comment]
@shenli @coocood PTAL

File: table/tables/index.go [hunk 2]
```diff
@@ -100,7 +100,9 @@ type index struct {
 	idxInfo *model.IndexInfo
 	prefix  kv.Key
 
-	buffer []byte // It's used reduce the number of new slice when multiple index keys are created.
+	// buffer is used reduce the number of new slice when multiple index keys are created.
+	// It is not thread-safe, is only used in ddl index reorganation stage, please don't use it in other case.
```
Comment (shenli) [inline table/tables/index.go]: index reorganization

File: table/tables/index.go [hunk 3]
```diff
@@ -159,12 +173,7 @@ func (c *index) GenIndexKey(sc *stmtctx.StatementContext, indexedValues []types.
 			}
 		}
 	}
-
-	if c.buffer != nil {
-		key = c.buffer[:0]
-	} else {
-		key = make([]byte, 0, len(c.prefix)+len(indexedValues)*9+9)
-	}
+	key = c.getIndexKeyBuf(buf, len(c.prefix)+len(indexedValues)*9+9)
```
Comment (shenli) [inline table/tables/index.go]: This function is not only used in DDL. So is it safe to use the c.buffer in this function?

Comment (winkyao) [reply table/tables/index.go]: This function is not only used in DDL, but c.buffer is only set in DDL. so c.buffer will not be used by others.

File: table/tables/index.go [hunk 4]
```diff
@@ -301,8 +313,12 @@ func (c *index) Exist(sc *stmtctx.StatementContext, rm kv.RetrieverMutator, inde
 	return true, h, nil
 }
 
-func (c *index) FetchValues(r []types.Datum) ([]types.Datum, error) {
-	vals := make([]types.Datum, len(c.idxInfo.Columns))
+func (c *index) FetchValues(r []types.Datum, vals []types.Datum) ([]types.Datum, error) {
+	needLength := len(c.idxInfo.Columns)
+	if vals == nil || cap(vals) < needLength {
```
Comment (tiancaiamao) [inline table/tables/index.go]:
```
if len(vals)  < needLength {
    vals = make([]types.Datum, needLength)
}
```

if 
cap(vals) = 100
len(vals) = 1
needLength = 10
line 321 will panic.

Comment (winkyao) [reply table/tables/index.go]:
```
package main

import "fmt"

func main() {
	vals := make([]int, 1, 100)
	val := vals[:10]
	fmt.Println(val)
}
```
Not panic...

Comment (tiancaiamao) [reply table/tables/index.go]: Wow, I did not expect Go allow this feature.

Comment (winkyao) [reply table/tables/index.go]: Yes, golang support it.

**Comment by coocood** [comment]
LGTM

**Comment by shenli** [comment]
LGTM

**Pull request merged**